### PR TITLE
[DHIS2-5963] Allow access to metadata when item is not DataShareable

### PR DESCRIPTION
**Pull request by luciano-fiandesio**
… app

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java [hunk 1]
```diff
@@ -94,30 +94,22 @@ public boolean isDataShareable( Class<?> klass )
...
-
+        
         if ( canAccess( user, schema.getAuthorityByType( AuthorityType.READ ) ) )
         {
             if ( object instanceof CategoryOptionCombo )
             {
                 return checkOptionComboSharingPermission( user, object, Permission.READ );
             }
 
-            if ( !schema.isShareable() || object.getUser() == null || object.getPublicAccess() == null || checkUser( user, object )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java]: What's the reason for removing `object.getUser() == null` check?

File: dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java [hunk 2]
```diff
@@ -736,4 +678,27 @@ private boolean checkOptionComboSharingPermission( User user, IdentifiableObject
...
+        if ( object == null || haveOverrideAuthority( user ) )
+        {
+            return true;
+        }
+
+        return schemaService.getSchema( object.getClass() ) == null;
+
+    }
+
+    private boolean commonWriteCheck(Schema schema, User user, IdentifiableObject object ) {
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java]: Rename to `writeCommonCheck` to be consistent with method above.

File: dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java (hunk 1 above)
Comment (luciano-fiandesio) [reply dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java]: The original `if` condition goes like:

`if ( !schema.isShareable() || object.getUser() == null || object.getPublicAccess() == null || checkUser( user, object )`

The method `checkUser(user, object)` internally checks for `object.getUser() == null` so I removed the redundant check (unless we want to keep it there to exit early).

File: dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java [hunk 3]
```diff
@@ -164,21 +151,21 @@ public boolean canDataRead( User user, IdentifiableObject object )
...
+    }
+
     @Override
     public boolean canWrite( User user, IdentifiableObject object )
     {
-        if ( object == null || haveOverrideAuthority( user ) )
-        {
-            return true;
-        }
+        if ( readWriteCommonCheck( user, object ) ) return true;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java]: Use brackets on separate lines.

File: dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java [hunk 4]
```diff
@@ -216,18 +194,10 @@ else if ( schema.isImplicitPrivateAuthority() && checkSharingAccess( user, objec
     @Override
     public boolean canDataWrite( User user, IdentifiableObject object )
     {
-        if ( object == null || haveOverrideAuthority( user ) )
-        {
-            return true;
-        }
+        if ( readWriteCommonCheck( user, object ) ) return true;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java]: Use brackets on separate lines.

File: dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java [hunk 5]
```diff
@@ -249,18 +219,10 @@ public boolean canDataWrite( User user, IdentifiableObject object )
     @Override
     public boolean canUpdate( User user, IdentifiableObject object )
     {
-        if ( object == null || haveOverrideAuthority( user ) )
-        {
-            return true;
-        }
+        if ( readWriteCommonCheck( user, object ) ) return true;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java]: Use brackets on separate lines.

File: dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java [hunk 6]
```diff
@@ -272,16 +234,7 @@ public boolean canUpdate( User user, IdentifiableObject object )
...
-            {
-                return true;
-            }
-
-            if ( checkSharingAccess( user, object ) &&
-                (checkUser( user, object ) || checkSharingPermission( user, object, Permission.WRITE )) )
-            {
-                return true;
-            }
+            return writeCommonCheck(schema, user, object);
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java]: Spaces, codestyle.

File: dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java [hunk 7]
```diff
@@ -295,18 +248,10 @@ else if ( schema.isImplicitPrivateAuthority() && checkSharingAccess( user, objec
     @Override
     public boolean canDelete( User user, IdentifiableObject object )
     {
-        if ( object == null || haveOverrideAuthority( user ) )
-        {
-            return true;
-        }
+        if ( readWriteCommonCheck( user, object ) ) return true;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java]: Use brackets on separate lines.

**Pull request merged**